src/exchange/providers/retry_with_back_off_decorator.py

```diff
@@ -0,0 +1,54 @@
+import time
+from functools import wraps
+from typing import Any, Callable, Dict, List, Optional
+
+from httpx import HTTPStatusError, Response
+
+
+def retry_with_backoff(
+        should_retry: Callable, 
+        max_retries: Optional[int] = 5, 
+        initial_wait: Optional[float] = 10, 
+        backoff_factor: Optional[float] = 1, 
+        handle_retry_exhausted: Optional[Callable] = None) ->  Callable:
+    def decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(*args: List, **kwargs: Dict) -> Any:  # noqa: ANN401
+            result = None
+            for retry in range(max_retries):
+                result = func(*args, **kwargs)
+                if not should_retry(result):
+                    return result
+                if (retry + 1) == max_retries:
+                    break
+                sleep_time = initial_wait + (backoff_factor * (2 ** retry))
+                time.sleep(sleep_time)
+            if handle_retry_exhausted:
+                handle_retry_exhausted(result, max_retries)
+            return result
+        return wrapper
+    return decorator
+
+def retry_httpx_request(
+        retry_on_status_code: List[int] = [429, 529, 500],
```

Comment:
Should we change this to `range(400, 999)` or something? (or change the logic)

Reply:
we could use a [range map](https://collections-extended.lenzm.net/range_map/)?

Reply:
Discussed on Slack; adds an extra dep so decided not to do it